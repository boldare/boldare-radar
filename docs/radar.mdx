---
sidebar_position: 1
hide_table_of_contents: true
---

import { RadarContainer } from "../src/components/Radar";

# Radar

<RadarContainer />
