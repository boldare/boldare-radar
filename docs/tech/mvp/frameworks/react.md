# React

## What is React?

[React](https://reactjs.org/) is a JavaScript library for building user interfaces

## 💡 Resources

### 📜 Articles

- [Documentation](https://reactjs.org/docs/getting-started.html)
- [Documentation V2](https://beta.reactjs.org/)
- [Handling state complexity](https://dev.to/davidkpiano/no-disabling-a-button-is-not-app-logic-598i)

### ✍️ Blogs

- [Overreacted](https://overreacted.io/)
- [Josh W Comeau about React](https://www.joshwcomeau.com/tutorials/react/)
