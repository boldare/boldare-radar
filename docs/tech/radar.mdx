---
sidebar_position: 1
hide_table_of_contents: true
---

import { RadarContainer } from "../../src/components/Radar";
import { RadarIntro } from "../../src/components/RadarIntro";

# Radar

<RadarIntro />

<RadarContainer />
