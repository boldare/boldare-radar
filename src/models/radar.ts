type QuadrantName =
  | "Techniques"
  | "Tools"
  | "Platforms & Services"
  | "Frameworks";

type RingName = "Prototype" | "MVP" | "PMF/Scaling";

interface Comment {
  author: string;
  content: string;
}

export interface Entry {
  quadrant: QuadrantName;
  ring: RingName;
  label: string;
  active: boolean;
  moved: number;
  links?: string[];
  comments?: Comment[];
}

export interface Ring {
  name: RingName;
  color: string;
}

export interface Quadrant {
  name: QuadrantName;
}
