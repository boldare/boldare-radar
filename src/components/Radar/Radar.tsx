import * as React from "react";
import { useRadarData } from "../../hooks/useRadarData";
import {
  Entry,
  Quadrant,
  QuadrantName,
  Ring,
  RingName,
} from "../../models/radar";
import { radar_visualization } from "../../radar/visualization";
import { useHistory } from "@docusaurus/router";

interface RadarProps {
  quadrants: Quadrant[];
  rings: Ring[];
  entries: Entry[];
  handleEntryClick: (entry: Entry) => void;
}

const quadrantMap: Record<QuadrantName, number> = {
  [QuadrantName.Frameworks]: 0,
  [QuadrantName.PlatformsAndServices]: 1,
  [QuadrantName.Techniques]: 2,
  [QuadrantName.Tools]: 3,
};

const ringMap: Record<RingName, number> = {
  [RingName.Prototype]: 0,
  [RingName.MVP]: 1,
  [RingName.Scaleup]: 2,
};

export function Radar({
  quadrants,
  entries,
  rings,
  handleEntryClick,
}: RadarProps) {
  const ref = React.useRef(null);
  const radarRendered = React.useRef(false);

  const formattedEntries = React.useMemo(
    () =>
      entries.map((entry) => ({
        ...entry,
        ring: ringMap[entry.ring],
        quadrant: quadrantMap[entry.quadrant],
      })),
    [entries]
  );

  React.useEffect(() => {
    if (radarRendered.current) {
      return;
    }

    radar_visualization({
      ref,
      width: 1450,
      height: 1000,
      colors: {
        background: "#fff",
        grid: "#bbb",
        inactive: "#ddd",
      },
      title: "Boldare Radar",
      quadrants,
      entries: formattedEntries,
      rings,
      print_layout: true,
      handleEntryClick,
      // zoomed_quadrant: 0,
    });

    radarRendered.current = true;
  }, [ref, radarRendered, entries, rings, quadrants]);

  return <svg ref={ref} />;
}

export function RadarContainer() {
  const data = useRadarData();
  const history = useHistory();

  return (
    <div style={{ overflow: "auto" }}>
      <Radar {...data} handleEntryClick={(entry) => history.push(entry.slug)} />
    </div>
  );
}
