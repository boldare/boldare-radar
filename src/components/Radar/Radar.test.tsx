import * as React from "react";
import { render } from "@testing-library/react";
import { Radar } from "./Radar";
import { Entry, Quadrant, Ring } from "../../models/radar";

const quadrants = [
  { name: "Techniques" },
  { name: "Tools" },
  { name: "Platforms & Services" },
  { name: "Frameworks" },
] as Quadrant[];

const entries = [
  {
    active: true,
    color: "#fbdb84",
    id: "27",
    index: 0,
    label: "AWS Athena",
    moved: 0,
    quadrant: "Platforms & Services",
    ring: "PMF/Scaling",
  },
  {
    quadrant: "Platforms & Services",
    ring: "PMF/Scaling",
    label: "Google Bigtable",
    active: true,
    moved: 0
  },
  {
    quadrant: "Tools",
    ring: "PMF/Scaling",
    label: "Airflow",
    active: true,
    moved: 0
  },
  {
    quadrant: "Platforms & Services",
    ring: "PMF/Scaling",
    label: "AWS DynamoDB",
    active: true,
    moved: 0
  },
  {
    quadrant: "Tools",
    ring: "PMF/Scaling",
    label: "Nginx",
    active: true,
    moved: 0,
  },
] as Entry[];

const rings = [
  {
    color: "#93c47d",
    name: "Prototype",
  },
  {
    color: "#93d2c2",
    name: "MVP",
  },
  {
    color: "#fbdb84",
    name: "PMF/Scaling",
  },
] as Ring[];

describe("Radar component", () => {
  it("renders correctly", () => {
    const { baseElement } = render(
      <Radar entries={entries} rings={rings} quadrants={quadrants} />
    );

    expect(baseElement).toMatchSnapshot();
  });
});
