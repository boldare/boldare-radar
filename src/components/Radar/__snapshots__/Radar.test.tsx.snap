// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radar component renders correctly 1`] = `
<body>
  <div>
    <svg
      height="1000"
      style="background-color: rgb(255, 255, 255);"
      width="1450"
    >
      <g
        transform="translate(725,500)"
      >
        <g>
          <line
            style="stroke: #bbb; stroke-width: 1;"
            x1="0"
            x2="0"
            y1="-400"
            y2="400"
          />
          <line
            style="stroke: #bbb; stroke-width: 1;"
            x1="-400"
            x2="400"
            y1="0"
            y2="0"
          />
          <defs>
            <filter
              height="1"
              id="solid"
              width="1"
              x="0"
              y="0"
            >
              <feflood
                flood-color="rgb(0, 0, 0, 0.8)"
              />
              <fecomposite
                in="SourceGraphic"
              />
            </filter>
          </defs>
          <circle
            cx="0"
            cy="0"
            r="130"
            style="fill: none; stroke: #bbb; stroke-width: 1;"
          />
          <text
            style="fill: #e5e5e5; font-family: Arial, Helvetica; font-size: 42px; font-weight: bold; pointer-events: none; user-select: none;"
            text-anchor="middle"
            y="-68"
          >
            Prototype
          </text>
          <circle
            cx="0"
            cy="0"
            r="220"
            style="fill: none; stroke: #bbb; stroke-width: 1;"
          />
          <text
            style="fill: #e5e5e5; font-family: Arial, Helvetica; font-size: 42px; font-weight: bold; pointer-events: none; user-select: none;"
            text-anchor="middle"
            y="-158"
          >
            MVP
          </text>
          <circle
            cx="0"
            cy="0"
            r="310"
            style="fill: none; stroke: #bbb; stroke-width: 1;"
          />
          <text
            style="fill: #e5e5e5; font-family: Arial, Helvetica; font-size: 42px; font-weight: bold; pointer-events: none; user-select: none;"
            text-anchor="middle"
            y="-248"
          >
            PMF/Scaling
          </text>
        </g>
        <text
          style="font-family: Arial, Helvetica; font-size: 34px;"
          transform="translate(-675,-420)"
        >
          Boldare Radar
        </text>
        <text
          space="preserve"
          style="font-family: Arial, Helvetica; font-size: 10px;"
          transform="translate(-675,420)"
        >
          ▲ moved up     ▼ moved down
        </text>
        <g>
          <text
            style="font-family: Arial, Helvetica; font-size: 18px;"
            transform="translate(450,45)"
          >
            Techniques
          </text>
          <text
            style="font-family: Arial, Helvetica; font-size: 12px; font-weight: bold;"
            transform="translate(450,74)"
          >
            Prototype
          </text>
          <text
            style="font-family: Arial, Helvetica; font-size: 12px; font-weight: bold;"
            transform="translate(450,110)"
          >
            MVP
          </text>
          <text
            style="font-family: Arial, Helvetica; font-size: 12px; font-weight: bold;"
            transform="translate(570,74)"
          >
            PMF/Scaling
          </text>
          <text
            style="font-family: Arial, Helvetica; font-size: 18px;"
            transform="translate(-675,45)"
          >
            Tools
          </text>
          <text
            style="font-family: Arial, Helvetica; font-size: 12px; font-weight: bold;"
            transform="translate(-675,74)"
          >
            Prototype
          </text>
          <text
            style="font-family: Arial, Helvetica; font-size: 12px; font-weight: bold;"
            transform="translate(-675,110)"
          >
            MVP
          </text>
          <text
            style="font-family: Arial, Helvetica; font-size: 12px; font-weight: bold;"
            transform="translate(-555,74)"
          >
            PMF/Scaling
          </text>
          <a
            href="#"
          >
            <text
              class="legend12"
              id="legendItem4"
              style="font-family: Arial, Helvetica; font-size: 11px;"
              transform="translate(-555,90)"
            >
              4. Airflow
            </text>
          </a>
          <a
            href="#"
          >
            <text
              class="legend12"
              id="legendItem5"
              style="font-family: Arial, Helvetica; font-size: 11px;"
              transform="translate(-555,102)"
            >
              5. Nginx
            </text>
          </a>
          <text
            style="font-family: Arial, Helvetica; font-size: 18px;"
            transform="translate(-675,-355)"
          >
            Platforms & Services
          </text>
          <text
            style="font-family: Arial, Helvetica; font-size: 12px; font-weight: bold;"
            transform="translate(-675,-326)"
          >
            Prototype
          </text>
          <text
            style="font-family: Arial, Helvetica; font-size: 12px; font-weight: bold;"
            transform="translate(-675,-290)"
          >
            MVP
          </text>
          <text
            style="font-family: Arial, Helvetica; font-size: 12px; font-weight: bold;"
            transform="translate(-555,-326)"
          >
            PMF/Scaling
          </text>
          <a
            href="#"
          >
            <text
              class="legend22"
              id="legendItem1"
              style="font-family: Arial, Helvetica; font-size: 11px;"
              transform="translate(-555,-310)"
            >
              1. AWS Athena
            </text>
          </a>
          <a
            href="#"
          >
            <text
              class="legend22"
              id="legendItem2"
              style="font-family: Arial, Helvetica; font-size: 11px;"
              transform="translate(-555,-298)"
            >
              2. AWS DynamoDB
            </text>
          </a>
          <a
            href="#"
          >
            <text
              class="legend22"
              id="legendItem3"
              style="font-family: Arial, Helvetica; font-size: 11px;"
              transform="translate(-555,-286)"
            >
              3. Google Bigtable
            </text>
          </a>
          <text
            style="font-family: Arial, Helvetica; font-size: 18px;"
            transform="translate(450,-355)"
          >
            Frameworks
          </text>
          <text
            style="font-family: Arial, Helvetica; font-size: 12px; font-weight: bold;"
            transform="translate(450,-326)"
          >
            Prototype
          </text>
          <text
            style="font-family: Arial, Helvetica; font-size: 12px; font-weight: bold;"
            transform="translate(450,-290)"
          >
            MVP
          </text>
          <text
            style="font-family: Arial, Helvetica; font-size: 12px; font-weight: bold;"
            transform="translate(570,-326)"
          >
            PMF/Scaling
          </text>
        </g>
        <g
          id="rink"
        >
          <g
            class="blip"
            transform="translate(-555,-310)"
          >
            <circle
              fill="#fbdb84"
              r="9"
            />
            <text
              style="fill: #fff; font-family: Arial, Helvetica; font-size: 9px; pointer-events: none; user-select: none;"
              text-anchor="middle"
              y="3"
            >
              1
            </text>
          </g>
          <g
            class="blip"
            transform="translate(-555,-298)"
          >
            <circle
              fill="#fbdb84"
              r="9"
            />
            <text
              style="fill: #fff; font-family: Arial, Helvetica; font-size: 9px; pointer-events: none; user-select: none;"
              text-anchor="middle"
              y="3"
            >
              3
            </text>
          </g>
          <g
            class="blip"
            transform="translate(-555,114)"
          >
            <circle
              fill="#fbdb84"
              r="9"
            />
            <text
              style="fill: #fff; font-family: Arial, Helvetica; font-size: 9px; pointer-events: none; user-select: none;"
              text-anchor="middle"
              y="3"
            >
              4
            </text>
          </g>
          <g
            class="blip"
            transform="translate(-555,-274)"
          >
            <circle
              fill="#fbdb84"
              r="9"
            />
            <text
              style="fill: #fff; font-family: Arial, Helvetica; font-size: 9px; pointer-events: none; user-select: none;"
              text-anchor="middle"
              y="3"
            >
              2
            </text>
          </g>
          <g
            class="blip"
            transform="translate(-555,138)"
          >
            <circle
              fill="#fbdb84"
              r="9"
            />
            <text
              style="fill: #fff; font-family: Arial, Helvetica; font-size: 9px; pointer-events: none; user-select: none;"
              text-anchor="middle"
              y="3"
            >
              5
            </text>
          </g>
        </g>
        <g
          id="bubble"
          style="opacity: 0; pointer-events: none; user-select: none;"
          x="0"
          y="0"
        >
          <rect
            rx="4"
            ry="4"
            style="fill: #333;"
          />
          <text
            style="font-family: sans-serif; font-size: 10px; fill: #fff;"
          />
          <path
            d="M 0,0 10,0 5,8 z"
            style="fill: #333;"
          />
        </g>
      </g>
    </svg>
  </div>
</body>
`;
